File: 11/code_one.py
#!/usr/bin/env python3.9
import copy

def count_neighbours(r, c, seats):
    occupied = 0
    if r > 0 and c > 0 and seats[r-1][c-1] == '#':
        occupied += 1
    if r > 0 and seats[r-1][c] == '#':
        occupied += 1
    if r > 0 and c < len(seats[r]) - 1 and seats[r-1][c+1] == '#':
        occupied += 1
    if c > 0 and seats[r][c-1] == '#':
        occupied += 1
    if c < len(seats[r]) - 1 and seats[r][c+1] == '#':
        occupied += 1
    if r < len(seats) - 1 and c > 0 and seats[r+1][c-1] == '#':
        occupied += 1
    if r < len(seats) - 1 and seats[r+1][c] == '#':
        occupied += 1
    if r < len(seats) - 1 and c < len(seats[r]) - 1 and seats[r+1][c+1] == '#':
        occupied += 1
    return occupied

def iterate_seats(seats):
    new_seats = copy.deepcopy(seats)
    for r, row in enumerate(seats):
        for c, col in enumerate(row):
            if col in '#L':
                cnt = count_neighbours(r, c, seats)
                if col == 'L' and cnt == 0:
                    new_seats[r][c] = '#'
                elif col == '#' and cnt >= 4:
                    new_seats[r][c] = 'L'
    return new_seats

# Read layout
seats = []
with open('in.txt', 'r') as fd:
    for line in fd:
        seats.append([])
        for c in line[:-1]:
            seats[-1].append(c)

old_seats = []
while seats != old_seats:
    old_seats = copy.deepcopy(seats)
    seats = iterate_seats(old_seats)
    #for row in seats:
    #    print(row)
    #print()

cnt = 0
for row in seats:
    for col in row:
        if col == '#':
            cnt += 1
print(f"{cnt} seats are occupied")

File: 11/code_one.cpp
//
// Advent of Code 2020
//       Day 11
//
// © Ralph Ganszky
//

#include <filesystem>
#include <fstream>
#include <iostream>
#include <numeric>
#include <string>
#include <vector>

enum class Seat : int8_t { Empty, Free, Occupied };

int main(int argc, char* argv[]) {
    // Determine input filename - default in.txt
    std::string in_name = "in.txt";
    if (argc > 1) {
        if (std::filesystem::exists(argv[1])) {
            in_name = argv[1];
        }
    }
    std::filesystem::path input_path(in_name);
    std::ifstream in(input_path);
    if (!in.is_open())
        throw std::runtime_error(std::string("Can't open ") + input_path.string() +
                                 " for reading!");

    // Read layout
    std::vector<std::vector<Seat>> seats;
    std::string line;
    while (std::getline(in, line)) {
        seats.emplace_back(std::vector<Seat>);
        seats.back().reserve(line.size());
        std::transform(
            line.begin(), line.end(), seats.back().begin(), [&](const char a) -> Seat {
                if (a == '.')
                    return Seat::Empty;
                else if (a == 'L')
                    return Seat::Free;
                else if (a == '#')
                    return Seat::Occupied;
                else
                    throw std::runtime_error(std::string("Unknown Seat code ") + a + "!");
            });
    }

    auto count{0LL};

    std::cout << count << " seats are occupied\n";

    return 0;
}
